### PR TITLE
Add peval macro.

### PR DESCRIPTION
**Pull request by aadcg**
A simpler macro than pflet that covers common use cases.

Do you find it useful? @ambrevar @aartaka @jmercouris @pdelfino.

### CODE REVIEW

File: source/mode/web.lisp [hunk 1]
```diff
@@ -255,15 +255,11 @@ define which elements are picked up by element hinting.")
...
-            (ps:chain document (query-selector-all "[rel=next]") 0 (click))))
-    (go-next)))
+  (plambda (ps:chain document (query-selector-all "[rel=next]") 0 (click))))
 
 (define-command go-previous ()
   "Navigate to the previous element according to the HTML 'rel' attribute."
-  (pflet ((go-previous ()
-            (ps:chain document (query-selector-all "[rel=prev]") 0 (click))))
-    (go-previous)))
+  (plambda (ps:chain document (query-selector-all "[rel=prev]") 0 (click))))
```
Comment (aartaka) [inline source/mode/web.lisp:262]: We have `nyxt/ps:qsa` for that.

File: source/mode/reading-line.lisp [hunk 2]
```diff
@@ -45,35 +45,31 @@ mode."
 
 (define-command jump-to-reading-line-cursor (&key (buffer (current-buffer)))
   "Move the view port to show the reading line cursor."
-  (pflet ((jump-to-cursor ()
-                          (ps:chain (ps:chain document (query-selector "#reading-line-cursor")) (scroll-into-view-if-needed))))
-    (with-current-buffer buffer
-      (jump-to-cursor))))
+  (with-current-buffer buffer
+    (plambda (ps:chain
+              (ps:chain document (query-selector "#reading-line-cursor"))
```
Comment (aartaka) [inline source/mode/reading-line.lisp:50]: I believer you can remove the inner `ps:chain` there -- the outer one will work just fine.

File: source/mode/element-hint.lisp [hunk 3]
```diff
@@ -446,14 +446,12 @@ visible nosave active buffer."
...
-            (ps:dolist (element (nyxt/ps:qsa document ".nyxt-hint"))
-              (if (or (= (ps:chain element style opacity) "1")
-                      (= (ps:chain element style opacity) ""))
-                  (setf (ps:chain element style opacity) "0.2")
-                  (setf (ps:chain element style opacity) "1.0")))))
-    (with-current-buffer buffer
-      (toggle-transparent))))
+  (with-current-buffer buffer
+    (plambda (ps:dolist (element (nyxt/ps:qsa document ".nyxt-hint"))
+               (if (or (= (ps:chain element style opacity) "1")
```
Comment (aartaka) [inline source/mode/element-hint.lisp:451]: `ps:@` will be shorter than `ps:chain` here.

**Review by aartaka — COMMENTED** [review]
That's a much welcome change, I like it!

I've added some things that it would be cool to fix while you're working with our Parenscript. You fine with fixing those?

File: source/renderer-script.lisp [hunk 4]
```diff
@@ -24,6 +24,10 @@ The function can be passed ARGS."
                   collect (transform-definition name lambda-list body))
        ,@body)))
 
+(export-always 'plambda)
+(defmacro plambda (&body body)
```
[the remark this answers is not in the document]
Comment (aadcg) [reply source/renderer-script.lisp:28]: That's a good point indeed. I wondered if it should semantically be closer to lambda, but then we'd have `pflet`.

File: source/mode/web.lisp (hunk 1 above)
Comment (aadcg) [reply source/mode/web.lisp:262]: I believe you'd be correct if instead of `(ps:chain document (query-selector-all "[rel=next]") 0 (click))`, it'd be `(ps:chain document (query-selector-all "[rel=next]"))`. 

Nonetheless, I know nothing about JS :D

**Comment by aadcg** [comment]
@aartaka, all addressed. Care to have a look again?

**Comment by jmercouris** [comment]
Fascinating. I seem to remember some gotcha with `ffi-buffer-evaluate-javascript`. I had gone down a long rabbit hole of why we can't actually return the results of a javascript evaluation inline. Unfortunately I don't remember. @Ambrevar do you remember? It had something to do with calispel if I recall, and how  value would never be ready or something.

If it works though, looks a welcome change to me :-). Thanks Andre!

File: source/mode/web.lisp (hunk 1 above)
Comment (aartaka) [reply source/mode/web.lisp]: You can nest forms in JS, can you not? Thus `(ps:chain (nyxt/ps:qsa "rel=next") 0 (click))` should work and will be much shorter.

**Comment by aartaka** [comment]
@jmercouris, it was because of renderer/non-renderer thread duality, and it's long fixed, iirc.

File: source/mode/element-frame.lisp [hunk 5]
```diff
@@ -149,10 +149,9 @@
 
 (defun frame-element-clear ()
   "Clear the selection frame created by the user."
-  (pflet ((remove-overlay ()
-            (ps:chain document (get-element-by-id "nyxt-rectangle-selection") (remove))
-            (ps:chain document (get-element-by-id "nyxt-overlay") (remove))))
-    (remove-overlay)))
+  (peval
+    (ps:chain document (get-element-by-id "nyxt-rectangle-selection") (remove))
```
Comment (aartaka) [inline source/mode/element-frame.lisp:153]: I wonder if `nyxt/ps:qs` can shorten this too... Should it?

File: source/mode/plaintext-editor.lisp [hunk 6]
```diff
@@ -41,8 +41,7 @@
 
 (defmethod get-content ((editor plaintext-editor-mode))
   (with-current-buffer (buffer editor)
-    (pflet ((get-content () (ps:chain document (query-selector "#editor") value)))
-      (get-content))))
+    (peval (ps:chain document (query-selector "#editor") value))))
```
Comment (aartaka) [inline source/mode/plaintext-editor.lisp:44]: A good place for `nyxt/ps:qs` too :)

File: source/mode/reading-line.lisp [hunk 7]
```diff
@@ -45,35 +45,30 @@ mode."
 
 (define-command jump-to-reading-line-cursor (&key (buffer (current-buffer)))
   "Move the view port to show the reading line cursor."
-  (pflet ((jump-to-cursor ()
-                          (ps:chain (ps:chain document (query-selector "#reading-line-cursor")) (scroll-into-view-if-needed))))
-    (with-current-buffer buffer
-      (jump-to-cursor))))
+  (with-current-buffer buffer
+    (peval (ps:chain document (query-selector "#reading-line-cursor"))
```
Comment (aartaka) [inline source/mode/reading-line.lisp:49]: `nyxt/ps:qs`

File: source/mode/reading-line.lisp [hunk 8]
```diff
@@ -45,35 +45,30 @@ mode."
...
-      (setf (ps:@ (ps:chain document (query-selector "#reading-line-cursor")) style top)
-            (+ (- original-position (ps:lisp step-size)) "px")))))
-    (with-current-buffer buffer
-      (cursor-up)))
+  (with-current-buffer buffer
+    (peval
+      (let ((original-position
+              (ps:chain (parse-int
+                         (ps:@ (ps:chain document (query-selector "#reading-line-cursor")) style top) 10))))
+        (setf (ps:@ (ps:chain document (query-selector "#reading-line-cursor")) style top)
```
Comment (aartaka) [inline source/mode/reading-line.lisp:60]: `nyxt/ps:qs`

File: source/mode/reading-line.lisp [hunk 9]
```diff
@@ -45,35 +45,30 @@ mode."
...
-      (setf (ps:@ (ps:chain document (query-selector "#reading-line-cursor")) style top)
-            (+ (+ original-position (ps:lisp step-size)) "px")))))
-    (with-current-buffer buffer
-      (cursor-down)))
+  "Move the reading line cursor down. If scrolling off screen, move the screen as well."
+  (with-current-buffer buffer
+    (peval (let ((original-position
+                     (ps:chain (parse-int
+                                (ps:@ (ps:chain document (query-selector "#reading-line-cursor")) style top) 10))))
+               (setf (ps:@ (ps:chain document (query-selector "#reading-line-cursor")) style top)
```
Comment (aartaka) [inline source/mode/reading-line.lisp:70]: `nyxt/ps:qs`

File: source/mode/web.lisp [hunk 10]
```diff
@@ -255,15 +255,11 @@ define which elements are picked up by element hinting.")
 
 (define-command go-next ()
   "Navigate to the next element according to the HTML 'rel' attribute."
-  (pflet ((go-next ()
-            (ps:chain document (query-selector-all "[rel=next]") 0 (click))))
-    (go-next)))
+  (peval (ps:chain document (query-selector-all "[rel=next]") 0 (click))))
```
Comment (aartaka) [inline source/mode/web.lisp:258]: `nyxt/ps:qsa`

File: source/mode/web.lisp [hunk 11]
```diff
@@ -255,15 +255,11 @@ define which elements are picked up by element hinting.")
...
 
 (define-command go-previous ()
   "Navigate to the previous element according to the HTML 'rel' attribute."
-  (pflet ((go-previous ()
-            (ps:chain document (query-selector-all "[rel=prev]") 0 (click))))
-    (go-previous)))
+  (peval (ps:chain document (query-selector-all "[rel=prev]") 0 (click))))
```
Comment (aartaka) [inline source/mode/web.lisp:262]: `nyxt/ps:qsa`

**Review by aartaka — COMMENTED** [review]
At this point, I'm just getting picky about `query-selector(-all)` used instead of `nyxt/ps:qs(a)`.

@aadcg, these comments are rather a recommendation than obligation. It you're bored with this change — we can merge it as is, it will harm no one ;)

File: source/buffer.lisp [hunk 12]
```diff
@@ -775,17 +775,17 @@ Delete it with `ffi-buffer-delete'"))
...
+        (slot-value buffer 'document-model)
+        (with-current-buffer buffer
+          (let ((value (slot-value buffer 'document-model))
+                (element-count (%count-dom-elements)))
+            (if (and value element-count
+                     ;; Check whether the difference in element count is significant.
+                     (< (abs (- (length (clss:select "*" value)) (truncate element-count)))
+                        (document-model-delta-threshold buffer)))
+                value
+                (update-document-model :buffer buffer)))))))
```
Comment (Ambrevar) [inline source/buffer.lisp:788]: Make sure to fix indentation in separate commits, to make code reviews easier.

**Comment by Ambrevar** [comment]
Looks good to me!

File: source/mode/web.lisp (hunk 1 above)
Comment (aadcg) [reply source/mode/web.lisp]: Oh, sure. It makes sense now.

File: source/buffer.lisp (hunk 12 above)
Comment (aadcg) [reply source/buffer.lisp:788]: Ups, will take it into account next time!

File: source/mode/element-frame.lisp (hunk 5 above)
Comment (aadcg) [reply source/mode/element-frame.lisp:153]: I'll leave this investigation for another time :p

**Comment by aadcg** [comment]
Looks ready for me.

I gave it a test with `reading-line-mode` and it's rocking.

**Comment by aartaka** [comment]
Thanks!

**Comment by aadcg** [comment]
@aartaka I think you did right, as I've committed to do it later at some point! Thanks.

**Comment by jmercouris** [comment]
> @jmercouris, it was because of renderer/non-renderer thread duality, and it's long fixed, iirc.

Right, yes! Thank you! How was it fixed? I did not think it's possible to fix on the case of when we are on a renderer thread?

**Comment by aartaka** [comment]
> > @jmercouris, it was because of renderer/non-renderer thread duality, and it's long fixed, iirc.
> 
> Right, yes! Thank you! How was it fixed? I did not think it's possible to fix on the case of when we are on a renderer thread?

I guess current code of `%within-renderer-thread` and `ffi-buffer-evaluate-javascript` will tell you more than I will be able to :)

**Comment by aartaka** [comment]
> @aartaka I think you did right, as I've committed to do it later at some point! Thanks.

SHould we open an issue about it?

**Comment by aadcg** [comment]
@aartaka, yes I'll do it!
